### ISSUE TITLE
Consistency of handling a wrong length of parameter

### ISSUE THREAD

**Issue by yutannihilation** [tidyverse/ggplot2#3026]
During reviewing #3024, I'm a bit confused to see multiple values are handled differently. I'm yet to find which behaviour is the supposed one.

``` r
library(ggplot2)

df <- data.frame(
  x = c(1,3,2,5),
  y = c("a","c","d","c")
)
```

`nudge_x` and `nudge_y` accept a different length of parameter w/ a warning.

``` r
p <- ggplot(df, aes(x, y)) +
  geom_point() +
  geom_label(aes(label = y),
             nudge_y = c(-0.1, 0.1, -0.1))

invisible(ggplot_build(p))
#> Warning in y + params$y: longer object length is not a multiple of shorter
#> object length
```

If the parameter is an aesthetic variable, it raises an error.

``` r
p <- ggplot(df, aes(x, y)) +
  geom_point() +
  geom_label(aes(label = y),
             colour = c("red", "red", "green"))
invisible(ggplot_build(p))
#> Error: Aesthetics must be either length 1 or the same as the data (4): colour
```

Some parameters just ignore multiple values (this seems because `gpar()` does so).

``` r
ggplot(df, aes(x, y)) +
  geom_point() +
  geom_label(aes(label = y),
             label.size = c(1, 10, 100))
```

![](https://i.imgur.com/nQ7CArm.png)

<sup>Created on 2018-12-06 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

**Comment by yutannihilation** [comment]
Basically, I agree with you!

> Is there a reason why it is done that way?

I don't see any reason for now, but, if I understand correctly, `position` is not allowed to require custom aesthetics currently. So, we might need some effort to change internals to make it possible. For example, this is `all` that are considered as aesthetics:

https://github.com/tidyverse/ggplot2/blob/4a607caeffcb52c686597b651d99fc0252ff8786/R/layer.r#L111

**Comment by clauswilke** [comment]
> if I understand correctly, position is not allowed to require custom aesthetics currently. 

That's correct. However, I was talking about something else. I was suggesting that `nudge_x` and `nudge_y` could be aesthetics of `geom_text()` and handled by its `draw_panel()` function directly. So nudging would be independent of the position adjustment, and in fact it would work with any position adjustment you may want to use.

**Comment by clauswilke** [comment]
I didn't mean to propose to remove `position_nudge()` either. :-) I'm just saying `geom_text()` doesn't need to use it. It could have its own nudging feature.

**Comment by clauswilke** [comment]
I made a PR (#3030) to explain what I mean. I see no technical problems. In either case, `position_nudge()` should be retained, because it is useful beyond `geom_text()`/`geom_label()`. But nudging is sufficiently common for those two that they may deserve more specialized treatment.
